src/blocks/networkmanager.rs

```diff
@@ -28,125 +32,371 @@ enum NetworkState {
-            "/org/freedesktop/NetworkManager",
-            "org.freedesktop.DBus.Properties",
-            "Get")
+    fn get(c: &Connection, path: Path, t: &str, property: &str) -> Result<Message> {
+        let m = Message::new_method_call("org.freedesktop.NetworkManager", path, "org.freedesktop.DBus.Properties", "Get")
             .block_error("networkmanager", "Failed to create message")?
-            .append2(
-                MessageItem::Str("org.freedesktop.NetworkManager".to_string()),
-                MessageItem::Str(property.to_string())
-            );
+            .append2(MessageItem::Str(t.to_string()), MessageItem::Str(property.to_string()));
 
         let r = c.send_with_reply_and_block(m, 1000);
 
         r.block_error("networkmanager", "Failed to retrieve property")
     }
 
+    fn get_property(c: &Connection, property: &str) -> Result<Message> {
+        Self::get(c, "/org/freedesktop/NetworkManager".into(), "org.freedesktop.NetworkManager", property)
+    }
+
     pub fn state(&self, c: &Connection) -> Result<NetworkState> {
-        let m = Self::get_property(c, "State")?;
+        let m = Self::get_property(c, "State").block_error("networkmanager", "Failed to retrieve state")?;
 
-        let state: Variant<u32> = m.get1()
-            .block_error("networkmanager", "Failed to read property")?;
+        let state: Variant<u32> = m.get1().block_error("networkmanager", "Failed to read property")?;
 
         Ok(NetworkState::from(state.0))
     }
 
-    pub fn connection_type(&self, c: &Connection) -> Result<ConnectionType> {
-        let m = Self::get_property(c, "PrimaryConnectionType")?;
+    pub fn primary_connection(&self, c: &Connection) -> Result<NMConnection> {
+        let m = Self::get_property(c, "PrimaryConnection").block_error("networkmanager", "Failed to retrieve primary connection")?;
+
+        let primary_connection: Variant<Path> = m.get1().block_error("networkmanager", "Failed to read primary connection")?;
+
+        if let Ok(conn) = primary_connection.0.as_cstr().to_str() {
+            if conn == "/" {
+                return Err(BlockError("networkmanager".to_string(), "No primary connection".to_string()));
+            }
+        }
+
+        Ok(NMConnection { path: primary_connection.0.clone() })
+    }
+
+    pub fn active_connections(&self, c: &Connection) -> Result<Vec<NMConnection>> {
+        let m = Self::get_property(c, "ActiveConnections").block_error("networkmanager", "Failed to retrieve active connections")?;
 
-        let connection_type: Variant<String> = m.get1()
-            .block_error("networkmanager", "Failed to read property")?;
+        let active_connections: Variant<Array<Path, Iter>> = m.get1().block_error("networkmanager", "Failed to read active connections")?;
 
-        Ok(ConnectionType::from(connection_type.0))
+        Ok(active_connections.0.into_iter().map(|x| NMConnection { path: x }).collect())
+    }
+}
+
+#[derive(Clone)]
+struct NMConnection<'a> {
```

Comment:
I think this should be `NmConnection`, in keeping with Rust's naming convention for types containing acronyms.

Reply:
Noted. Was stuck in Go mode, where all acryonyms are upper/lower-cased as a unit.